### ISSUE TITLE
In Gen 4, the combination Klutz + Iron Ball is not properly handled

### ISSUE THREAD

**Issue by UnderscorePoY** [smogon/damage-calc#560]
[Bulbapedia](https://bulbapedia.bulbagarden.net/wiki/Klutz_(Ability)#Generation_IV) and [this code comment](https://github.com/smogon/damage-calc/blob/master/calc/src/mechanics/gen4.ts#L147) agree that when a Pokémon with Klutz holds an Iron Ball, the speed drop should still be applied. This is not the case with the current implementation, as shown in this image :
![2023-07-12 15_31_40-Pokémon Damage Calculator — Mozilla Firefox](https://github.com/smogon/damage-calc/assets/71231361/b42112a2-b758-434d-a245-cf34742ddedc)

Possible fix : modify the `checkItem` [https://github.com/smogon/damage-calc/blob/master/calc/src/mechanics/util.ts#L192](here) to keep the item if the combo Klutz + Item Ball arises.

**Comment by shrianshChari** [comment]
What does it look like when you remove the Klutz ability?

**Comment by UnderscorePoY** [comment]
The Iron Ball speed drop is properly applied.

![2023-07-12 21_44_55-Pokémon Damage Calculator — Mozilla Firefox](https://github.com/smogon/damage-calc/assets/71231361/9fc69dc3-4b7f-4719-8d02-c6be57257302)

**Comment by UnderscorePoY** [comment]
I suspect the issue comes from the fact that `checkItem` gets rid of the item entirely if Klutz is the Ability. Since `checkItem` is common to all generations, one might want to keep track of the old item and compare with it [on this line](https://github.com/smogon/damage-calc/blob/master/calc/src/mechanics/gen4.ts#L148).

**Comment by shrianshChari** [comment]
That line appears to deal with the grounding effect of Iron Ball, which according to Bulbapedia, doesn't get ignored by Klutz. In other words, Klutz allows the speed to be halved still, but doesn't allow the Pokemon to be grounded. However, [this line](https://github.com/smogon/damage-calc/blob/b1384d62656df9c20f94b2e063f78357fb2b3849/calc/src/mechanics/util.ts#L126C28-L126C28) and [this line](https://github.com/smogon/damage-calc/blob/b1384d62656df9c20f94b2e063f78357fb2b3849/calc/src/mechanics/util.ts#L194) might have something to do with how the damage calc handles the speed of the Pokemon.

**Comment by UnderscorePoY** [comment]
You're right, i linked you to a completely wrong reference, sorry about this.
Your first link is the `checkItem` function that is common to all generations. Probably not being modified, due to possible side effects in the other calculators split per generation.
Your first link is indeed where the speed drop is supposed to be applied, but since `checkItem` was already called previously and discards the item even in the case of an Iron Ball, the item is gone. Thus the drop is not put inside `speedMods`.
